[ADD] stock_zipcube

gurneyalex opened the pull request:
Transitory module, real work done in #1121 

don't merge.

gurneyalex committed:
[ADD] stock_zipcube

Zipcube is a device produced by Bosche which allows weighing and measuring packages
in a way similar to Cubiscan devices.

This PR adds a stock_zipcube module which builds on the stock_cubiscan module to
add support for Zipcube devices.

Future plan:
* extract common bits in a base_stock_measurement_scale module to allow installing stock_zipcube without pulling the cubiscan dependencies

[IMP] stock_zipcube: show pop-up after scanning to press refresh

Before, after scanning with the zipcube, the measurements were
sent to Odoo but the wizard was not refreshed, so this was
confusing to the user. With this PR, a pop-up is shown to the
user that opened the wizard, warning he/she to press the button
that refreshes the screen.

Apart from that, some bugs have been corrected:
* Cubiscan-only fields were required by zipcube (device_address,
  port and timeout). So now they are required only if the device
  is cubiscan (and not zipcube). This has required moving the
  requirement attribute from the field (model) to the view.
* Pressing the button that sets the line that is going to receive
  the measures didn't set correctly the scanning device for the
  package of the line.
* The check for the ports is done only if the device is Cubiscan.
* Measure key max_weight is not supported, so has been changed in
  the code that reads it to weight.

simahawk commented on stock_zipcube/models/zipcube.py:
```diff
@@ -42,8 +42,26 @@ def _update_packaging_measures(self, measures):
+        # that you only notify a user which is the same than the one set in
+        # the environment.
+        with api.Environment.manage():
+            env = api.Environment(
```
can't you use `with_user`? :thinking:

sebalix commented on stock_zipcube/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,4 @@
+The first step is to configure the Packaging Types (Pallet, Box, ...) in Inventory > Configuration > Product Packaging Types.
+
+Configure the Zipcube device in Inventory > Configuration > Cubiscan Devices.
```
This menu will have to be renamed once the common module is created.

sebalix commented on stock_zipcube/controllers/main.py:
```diff
@@ -0,0 +1,59 @@
+        _logger.info("received %s", data)
+        self._check_secret(data["secret"])
+        data = self._convert_floats(data)
+        # convert the float values passed as strings to floats
```
I suppose this comment apply to the line above `data = self._convert_floats(data)`

sebalix commented on stock_zipcube/models/product_packaging.py:
```diff
@@ -0,0 +1,12 @@
+        "cubiscan.device",
+        string="Device which will scan the package",
+        help="Technical field set when an operator uses "
+        "the device to scan this package",
```
add `copy=False`?

sebalix commented on stock_zipcube/models/zipcube.py:
```diff
@@ -0,0 +1,67 @@
+_logger = logging.getLogger(__name__)
+
+
+class CubiscanDevice(models.Model):
```
```suggestion
class ZipcubeDevice(models.Model):
```

sebalix commented on stock_zipcube/wizard/cubiscan_wizard.py:
```diff
@@ -0,0 +1,29 @@
+from odoo import fields, models
+
+
+class CubiscanWizard(models.TransientModel):
```
```suggestion
class ZipcubeWizard(models.TransientModel):
```

sebalix commented on stock_zipcube/wizard/cubiscan_wizard.py:
```diff
@@ -0,0 +1,29 @@
+    device_driver = fields.Selection(related="device_id.driver")
+
+
+class CubiscanWizardLine(models.TransientModel):
```
same

simahawk commented on stock_zipcube/controllers/main.py:
```diff
@@ -0,0 +1,59 @@
+  --request POST \
+  --data '{"secret": "abcdefg", "barcode":"xyz", "weight": "12,3",'\
+  '"length": "123,1", "width": "456,5", "height": "789,2"}' \
+  https://integration.cosanum.odoo.camptocamp.ch/stock/zipcube/1/measurement
```
:fire:

Commit:
[IMP] stock_zipcube:

changes made because of the code review

carlosserratoro commented:
@sebalix & @simahawk changes were done.

carlosserratoro committed:
[IMP] stock_zipcube: secret check (#4)

* [IMP] stock_zipcube: secret check

The URL that accesses the Zipcube is now protected by
a secret key, stored in an environment variable. If the
secret provided doesn't match, an exception is raised
and the access if forbidden.


Co-authored-by: Alexandre Fayolle <alexandre.fayolle@camptocamp.com>

simahawk commented on stock_measuring_device/components/measuring_device_component.py:
```diff
@@ -0,0 +1,61 @@
+
+
+class MeasuringDevice(AbstractComponent):
+    _name = "measuring.device.component"
```
```suggestion
    _name = "measuring.device.[mixin|base]"
```

simahawk commented on stock_measuring_device/models/measuring_device.py:
```diff
@@ -0,0 +1,68 @@
+        with self.work_on(self._name) as work:
+            return work.component(usage=self.device_type)
+
+    def open_wizard(self):
```
the same method on the component: drop or forward to it?

gurneyalex replied:
drop on the component

simahawk commented on stock_measuring_device/models/measuring_device.py:
```diff
@@ -0,0 +1,68 @@
+        }
+
+    def _is_being_used(self):
+        with self.work_on(self._name) as work:
```
use _forward

gurneyalex replied:
instead of 'work' ?

simahawk commented on stock_measuring_device/models/product_packaging.py:
```diff
@@ -0,0 +1,25 @@
+        "to scan this package",
+    )
+
+    def _assign_measuring_device(self, device):
```
when I add such methods, normally I tend to prefix all of them w/ the same name because it eases introspection.
Eg: `_measuring_device_assign` and `_measuring_device_release`.
OTOH these methods could stay on the device, eg: `_assign_to(packaging)` and `_release_from(packaging)`.

simahawk commented on stock_measuring_device/wizard/measuring_wizard.py:
```diff
@@ -0,0 +1,183 @@
+    line_ids = fields.One2many("measuring.wizard.line", "wizard_id")
+    device_id = fields.Many2one("measuring.device", readonly=True)
+
+    # @api.model
```
to trash?

simahawk commented on stock_measuring_device/wizard/measuring_wizard.py:
```diff
@@ -0,0 +1,183 @@
+            "height": self.product_id.product_height,
+        }
+        product_dimension_uom = self.product_id.dimensional_uom_id
+        mm_uom = self.env.ref("stock_measuring_device.product_uom_mm")
```
wouldn't be better to set this uom on device model?

simahawk commented on stock_measuring_device/wizard/measuring_wizard.xml:
```diff
@@ -0,0 +1,111 @@
+                    />
+                </header>
+                <group>
+                    <group>
```
name at least inner group

simahawk commented on stock_measuring_device_zippcube/components/zippcube_device_component.py:
```diff
@@ -0,0 +1,54 @@
+        self.env["measuring.wizard.line"].flush(
+            ["packaging_id", "wizard_id", "scan_requested"]
+        )
+        self.env["measuring.wizard"].flush(["device_id"])
```
flush device_id for?

gurneyalex replied:
idk... code from Carlos. Will remove.

simahawk commented on stock_measuring_device_zippcube/components/zippcube_device_component.py:
```diff
@@ -0,0 +1,54 @@
+            ["packaging_id", "wizard_id", "scan_requested"]
+        )
+        self.env["measuring.wizard"].flush(["device_id"])
+        wizard_line = self.collection.env["measuring.wizard.line"].search(
```
the whole part about finding the line (line 26-38) should stay on the wizard IMO or on the device.
Something like `self.collection._get_line_to_scan()`.
Note: on the abstract component having a property `device` that return `self.collection` could make things more clear.

simahawk commented on stock_measuring_device_zippcube/controllers/main.py:
```diff
@@ -0,0 +1,79 @@
+        self._check_secret(data["secret"])
+        # convert the float values passed as strings to floats
+        data = self._convert_floats(data)
+        with device.work_on(device._name) as work:
```
the device model should have a method `_update_packaging_measures` which forwards the call.
In any case here you could use `_get_measuring_device` :wink:

simahawk commented on stock_measuring_device_zippcube/controllers/main.py:
```diff
@@ -0,0 +1,79 @@
+                _("No such Zippcube with name {}.").format(zippcube_device_name)
+            )
+
+        keys_missing = set(self.expected_keys) - set(data)
```
FTR this whole validation and converting could be achieved using Cerberus schema (as done by base_rest).

simahawk reviewed (commented):
@gurneyalex being alpha, no real blocker but it would nice to have some tests at least.

gurneyalex commented:
I moved this to #1121 to keep the old module until the refactoring is complete

gurneyalex committed:
remove dependency on web_notify

gurneyalex committed:
Revert "remove dependency on web_notify"

This reverts commit be5a0640a917fa846ea7383d83278e2523a4f4a7.
